Added GKE configs for test databases

tarzzz opened the pull request:
- [x] Tested

Part of https://github.com/plotly/streambed/issues/10022

tarzzz commented:
@scjody Please review..

The test failures are intermittent and unrelated to these Databases..

scjody commented on sample-storage/connections.yaml:
```diff
@@ -85,20 +85,20 @@
     database: plotly
     port: 50000
-    host: 35.184.35.183
+    host: 35.193.19.132
```
Can you add hostnames to `test.plotly.host` and use those instead of IP addresses?

scjody replied:
(I suggest setting a TTL of 60 seconds when you do so.)

scjody commented on test/docker/ibmdb2/README.md:
````diff
@@ -51,3 +51,25 @@ docker run -ti -p 50000:50000 pdc-db2
+To deploy to GKE, run the following command:
+
+```sh
+kubectl create -f db2-deployment.yaml
````
Please create files that can be applied using `kubectl apply`.

scjody commented on test/docker/ibmdb2/README.md:
````diff
@@ -51,3 +51,25 @@ docker run -ti -p 50000:50000 pdc-db2
+Then create a service using the following command:
+
+```sh
+kubectl create -f db2-service.yaml
````
Why not create the service right away?

scjody commented:
Note that this relates to https://github.com/plotly/streambed/issues/10022

@tarzzz As a reminder, please add any related issues to the top comment of any PR you make.

tarzzz commented:
@scjody: 

  - Moved all the configs to a `deployment/` directory.
  - Added instructions for creating a GKE cluster and running test databases in a README.
  - Created hostnames for test databases on `.test.plotly.host` and updated the code to use the same.
  - Checked that the configurations can be run properly using `kubectl apply`.

Tests are failing because of spark connections (some issue with connectivity to the test db). I am looking into the same..

scjody commented on sample-storage/connections.yaml:
```diff
@@ -85,20 +85,20 @@
     database: plotly
     port: 50000
-    host: 35.184.35.183
+    host: db2.test.plotly.host
```
👍

scjody commented on test/docker/README.md:
````diff
@@ -0,0 +1,25 @@
+To Create GKE Cluster:
+
+```sh
+gcloud container clusters create falcon-test-dbs  --zone us-central1-b --machine-type n1-standard-2 --num-nodes 3
````
Please add `--enable-autoupgrade` (this may be a `beta` command)

scjody commented on test/docker/README.md:
````diff
@@ -0,0 +1,25 @@
+To create deployments for test databases:
+
+```sh
+kubectl apply -f deployment/
````
👍

n-riesco commented on backend/settings.js:
```diff
@@ -59,7 +59,10 @@ const DEFAULT_SETTINGS = {
+    IS_RUNNING_INSIDE_ON_PREM: false,
+
+    // Whether it is running in Circle CI environment
+    IS_RUNNING_ON_CIRCLE_CI: false
```
No need for this. Please, use https://circleci.com/docs/2.0/env-vars/#circleci-environment-variable-descriptions

tarzzz replied:
Circle env vars cannot be used with `getSetting` because the environment variables used in falcon are of the `PLOTLY_CONNECTOR_*` format.

n-riesco replied:
There is no need to use `getSetting`, use `process.env.CIRCLECI` instead.

tarzzz replied:
I think we should be consistent in our usage. We **dont** use `process.env` anywhere in code for accessing environment variables.

Going forward if you intend to use `process.env` for accessing environment variables, please open a separate issue where we can get all environment variables use `process.env.*`.

n-riesco replied:
> I think we should be consistent in our usage. We dont use process.env anywhere in code for accessing environment variables.

That's not true.
`process.env` is used by both the backend and the frontend.
Note that it's also used by `webpack.config.*`.

`getSetting` is only used to access properties in `~/.plotly/connector/settings.yaml` and ENV variables starting with `PLOTLY_CONNECTOR`.

For simplicity's sake, please, implement this change as requested.

tarzzz replied:
No, We don't. We only use it in `webpack.config.*` because we cannot access `getSettings` there. Here is a grep excluding `webpack.*` and compiled bundles:

```
 ✝  ~/plotly/falcon-sql-client   gke-deployments±  ack 'process.env' -r .
app/store/configureStore.js
1:if (process.env.NODE_ENV === 'production') {

test/backend/settings.spec.js
16:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
21:        delete process.env.PLOTLY_CONNECTOR_STORAGE_PATH;
32:        process.env.PLOTLY_CONNECTOR_STORAGE_PATH = __dirname;
74:        process.env.PLOTLY_CONNECTOR_ACCESS_TOKEN_EXPIRY = String(now);

.eslintrc
55:        "no-process-env": [0],

webpack.config.headless.js
27:            'process.env.NODE_ENV': JSON.stringify('production')

backend/menus.js
86:            submenu: (process.env.NODE_ENV === 'development') ? [{
153:            submenu: (process.env.NODE_ENV === 'development') ? [{

backend/settings.js
158: * Load settings from process.env prefixed with `PLOTLY_CONNECTOR_`
165:    } else if (has(`PLOTLY_CONNECTOR_${settingName}`, process.env)) {
166:        const envObject = process.env[`PLOTLY_CONNECTOR_${settingName}`];
```

Note that we only check for `NODE_ENV`. The only other place is `settings.spec.js` which is used to modify how `getSetting` behaves in tests.

n-riesco replied:
> No, We don't.

The results you're quoting show that `process.env` is also used in `app/store/configureStore.js`, `backend/menus.js`. 


It's possible to use github to see that it's also used in `backend/main.development.js` and `backend/headless.development.js`. See https://github.com/plotly/falcon-sql-client/search?utf8=%E2%9C%93&q=process.env&type= :

![image](https://user-images.githubusercontent.com/6199391/34780284-a26f42f2-f61a-11e7-8bc7-09852f7a7985.png)

![image](https://user-images.githubusercontent.com/6199391/34780329-c72b5338-f61a-11e7-91a7-1de4bbab43b7.png)

![image](https://user-images.githubusercontent.com/6199391/34780366-df8ff424-f61a-11e7-896a-23c99573c33e.png)

![image](https://user-images.githubusercontent.com/6199391/34780451-19fb6d00-f61b-11e7-9f6a-dff9de440caa.png)

![image](https://user-images.githubusercontent.com/6199391/34780472-2899d158-f61b-11e7-8c78-01428c4aecc4.png)

n-riesco replied:
Anyway, as I explained above `getSetting` isn't meant to access the ENV variable `CIRCLECI`.

The ENV variable `CIRCLECI` is set automatically by CircleCI and we can access it simply by reading `process.env.CIRCLECI`.

What you're proposing here is that:
1. we add the seeting `IS_RUNNING_ON_CIRCLE_CI` to `backend/settings.js`
2. we set `PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` in `circle.yml`
3. and we read the value by calling `getSetting('IS_RUNNING_ON_CIRCLE_CI')`

Please, for the sake of simplicity, implement the requested change; i.e. just use `process.env.CIRCLECI` to read the value.

scjody replied:
Using `getSetting` looks like a clearer and more consistent solution than sprinkling `process.env` reads throughout the code. It's best to centralize configuration handling in one place with one mechanism, and `getSetting` does that.

I support reading `process.env.CIRCLECI` in order to determine if the setting should be set, but it sounds like there are good arguments for doing it the way @tarzzz has implemented it now, namely that it's consistent with the way the other settings are handled.

n-riesco replied:
@scjody I strongly disagree. `getSetting` isn't meant to access properties like 
 `process.env.CIRCLECI` (as I explained above).

scjody replied:
@n-riesco I don't see any technical reason why `getSetting` couldn't be enhanced to deal with `process.env.CIRCLECI` as well as environment variables beginning with `PLOTLY_CONNECTOR_`. I suggested it because if you don't want to add another environment variable then that provides a way forward while still using `getSetting`.

`getSetting` is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

Sprinkling `process.env.CIRCLECI` checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. (There is one `process.env` user in the backend as you pointed out, but it predates the existence of `getSetting` and should probably be changed for consistency. But that's a separate concern.)

n-riesco replied:
@scjody 

> I don't see any technical reason why getSetting couldn't be enhanced to deal with process.env.CIRCLECI as well as environment variables beginning with PLOTLY_CONNECTOR_.

Technically it's possible; from a design point of view, it is undesirable. I've posted a number of reasons and none of them hasn't been refuted yet (I'll list them again at this end of this comment).

> getSetting is the way we handle settings in the backend of this app. If you don't want to use it with `process.env.CIRCLECI` then let's just add another environment variable as @tarzzz is proposing.

The backend doesn't need to know about `process.env.CIRCLECI`, only the specs do.

> Sprinkling process.env.CIRCLECI checks throughout the code will result in something that's significantly less maintainable since you're adding a second configuration mechanism. 

As I showed in my previous comments, the code is already sprinkled with `process.env`. I agree with you that `process.env` should be removed from the code (but this will have to wait until I have a chance to work again on the `webpack.config.*.js`).

Anyway, this argument doesn't apply to `process.env.CIRCLECI`, because it'd be used only inside the tests.

---

To sum up, here's the list of arguments against introducing the setting `IS_RUNNING_ON_CIRCLE_CI`:
1. The simplest solution (that requires no changes in the code) is to use `process.env.CIRCLECI`. Introducing `IS_RUNNING_ON_CIRCLE_CI` requires changes in `backend/settings.js`.
2. `CIRCLECI` is set automatically by CircleCI  (`PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` would require to change `circle.yml`)
3. Falcon doesn't need to know about `process.env.CIRCLECI` (only the tests do).

n-riesco commented on test/docker/spark/Dockerfile:
```diff
@@ -2,7 +2,7 @@ FROM sequenceiq/hadoop-ubuntu:2.6.0
 # Get latest java:
 COPY bin/setup_java.sh /setup_java.sh
-RUN chmod +x /setup_java.sh ; /setup_java.sh
+RUN chmod +x /setup_java.sh; sync; /setup_java.sh;
```
please, document why `sync` is needed.

n-riesco replied:
Oh, I see this is an attempt to fix the failure in circleCI.

tarzzz commented:
@scjody : Made some updates on the PR..

  - Uploaded spark and IBM DB2 images publicly to quay.io [here](quay.io/plotly/falcon-test-db2) and [here](quay.io/plotly/falcon-test-spark)

  - IBM DB2 and Spark test databases run inside the circle containers. No more intermittent failures 🎉 
  - Updated spark dockerfile to tackle https://github.com/moby/moby/issues/9547
  - Developers who run test locally can still connect to GKE cluster, so that they dont have to run docker containers for spark and ibm db2

Please let me know if it looks good .. !! 😸

n-riesco commented on test/docker/spark/Dockerfile:
```diff
@@ -1,8 +1,11 @@
 # Get latest java:
 COPY bin/setup_java.sh /setup_java.sh
-RUN chmod +x /setup_java.sh ; /setup_java.sh
+RUN chmod +x /setup_java.sh; sync; /setup_java.sh;
```
My understanding of https://github.com/moby/moby/issues/9547#issuecomment-337684598 is that the `sync` workaround may fail randomly.

I think we should get rid of `chmod` and simply run `sh /setup_java.sh` or `bash /setup_java.sh`.

tarzzz replied:
We can't get rid of a command which is necessary.

n-riesco replied:
Why is it necessary?

tarzzz replied:
because `sh setup_java.sh` will also not work without changing the mode (as I recall, I have already tried those options while writing the Dockerfile). 

> My understanding of moby/moby#9547 (comment) is that the sync workaround may fail randomly.

I am pretty positive it will work (for our use case). We can always get back on it if it causes us issues in future.

n-riesco replied:
Why do you think it will work (despite the explanation in https://github.com/moby/moby/issues/9547#issuecomment-337684598)?

scjody replied:
I suggest trying the workaround in https://github.com/moby/moby/issues/9547#issuecomment-337684598 . It's more likely to work, and the downside (that we create an extra layer) isn't a big deal for test systems.

If that doesn't work for some reason let's stick with `sync` since it's better than nothing. It's unfortunate how terrible Docker is but that's something we need to deal with.

n-riesco commented on test/docker/spark/Dockerfile:
```diff
@@ -20,7 +23,7 @@ COPY conf/my.cnf /etc/
 # Install mysql as hive metastore
 COPY setup_mysql.sh /setup_mysql.sh
-RUN chmod +x /setup_mysql.sh && /setup_mysql.sh;
+RUN chmod +x /setup_mysql.sh; sync; /setup_mysql.sh;
```
ditto

n-riesco commented on test/docker/spark/Dockerfile:
```diff
@@ -39,12 +42,12 @@ COPY conf/* $SPARK_HOME/conf/
 
 # Setup Livy
-RUN chmod +x /bin/setup_livy.sh; /bin/setup_livy.sh;
+RUN chmod +x /bin/setup_livy.sh; sync; /bin/setup_livy.sh;
```
ditto

n-riesco commented on test/docker/spark/Dockerfile:
```diff
@@ -39,12 +42,12 @@ COPY conf/* $SPARK_HOME/conf/
 EXPOSE 8998
 
-ENTRYPOINT /etc/bootstrap.sh && service mysql start && chmod +x /bin/setup_database.sh && /bin/setup_database.sh && /livy/bin/livy-server;
+ENTRYPOINT /etc/bootstrap.sh && service mysql start && chmod +x /bin/setup_database.sh && sync && /bin/setup_database.sh && /livy/bin/livy-server;
```
ditto

n-riesco commented:
TODO
- [ ] update the scripts `docker:db2:build` and `docker:db2:start`
- [ ] add similar scripts for spark

I've noticed there is Dockerfile for impala but this PR doesn't include use it in CircleCI.
Is this on purpose?

n-riesco commented:
I've tried to use the spark DB at spark.test.plotly.host but it fails. See screenshot:

![image](https://user-images.githubusercontent.com/6199391/34769400-98eab204-f5f5-11e7-88bc-6321369c4230.png)

n-riesco commented:
The DB2 sample server at db2.test.plotly.host works for me:

![image](https://user-images.githubusercontent.com/6199391/34769588-2409bd94-f5f6-11e7-9012-c07ffcbd73a2.png)

tarzzz commented:
> update the scripts docker:db2:build and docker:db2:start

The script does not need to be updated as we are not updating the `Dockerfile` for DB2. Please let me know in case I am missing something.

>  add similar scripts for spark

We don't need it because users or developers are not required to run these docker containers on their machine.

tarzzz commented:
> I've noticed there is Dockerfile for impala but this PR doesn't include use it in CircleCI.
Is this on purpose?

I could not get it working on circle due to some issues. Since it is not a blocker for this PR (I only added  spark and db2 to get tests passing), I will create a separate issue to handle this. Moreover Impala tests do not fail intermittently so we should be good meanwhile.

n-riesco commented:
@tarzzz 

> The script does not need to be updated as we are not updating the Dockerfile for DB2. Please let me know in case I am missing something.

NVM, I'll do it myself.

> We don't need it because users or developers are not required to run these docker containers on their machine.

Once those sample DB go away, developers will have to run those docker containers. Note that the spark sample DB is failing at  at spark.test.plotly.host the moment.

> I could not get it working on circle due to some issues. Since it is not a blocker for this PR (I only added spark and db2 to get tests passing), I will create a separate issue to handle this. Moreover Impala tests do not fail intermittently so we should be good meanwhile.

Note that we've had to disable tests in datastores.routes.spec.js for impala.

n-riesco commented:
The impala sample server at impala.test.plotly.host works for me:

![image](https://user-images.githubusercontent.com/6199391/34774783-a1b36634-f608-11e7-87c1-1f0132c8f4d0.png)

scjody commented:
@tarzzz Thanks for the changes. This looks good to me now but you're going to have to work out the `PLOTLY_CONNECTOR_IS_RUNNING_ON_CIRCLE_CI` issue with @n-riesco, and it would be good to try the other workaround for the Docker issue.

tarzzz commented:
Updates made:

  - Updated Spark Dockerfile as per the comments.
  - Updated code to use `process.env` instead of `getSetting`
  - Updated GKE cluster to use 3 nodes instead of 2. 

After updating the nodes to 3, spark connector is working for me:

![image](https://user-images.githubusercontent.com/1546532/34811229-e0482d72-f6c5-11e7-98be-7000ee4bfc72.png)

@n-riesco : Please let me know if it looks okay.. !!

n-riesco commented:
@tarzzz I can't test it today, but LGTM. :dancer:

n-riesco commented:
Just checked the spark server and it works for me too (although it takes a very long time to connect).
